react/features/toolbox/components/AbstractToolboxItem.js

```diff
@@ -0,0 +1,192 @@
+// @flow
+
+import { Component } from 'react';
+
+type Styles = {
+
+    /**
+     * Style for the item's icon.
+     */
+    iconStyle: Object,
+
+    /**
+     * Style for the item itself.
+     */
+    style: Object,
+
+    /**
+     * Color for the item underlay (shows when clicked).
+     */
+    underlayColor: string
+};
+
+export type Props = {
+
+    /**
+     * A succinct description of what the item does. Used by accessibility
+     * tools and torture tests.
+     */
+    accessibilityLabel: string,
+
+    /**
+     * Whether this item is disabled or not. When disabled, clicking an the item
+     * has no effect, and it may reflect on its style.
+     */
+    disabled: boolean,
+
+    /**
+     * The name of the icon of this {@code ToolboxItem}.
+     */
+    iconName: string,
+
+    /**
+     * The text associated with this item. When `showLabel` is set to
+     * {@code true}, it will be displayed alongside the icon.
+     */
+    label: string,
+
+    /**
+     * On click handler.
+     */
+    onClick: Function,
+
+    /**
+     * Whether to show the label or not.
+     */
+    showLabel: boolean,
+
+    /**
+     * Collection of styles for the item. Used only on native.
+     */
+    styles: Styles,
+
+    /**
+     * Invoked to obtain translated strings.
+     */
+    t: ?Function,
+
+    /**
+     * The text to display in the tooltip. Used only on web.
+     */
+    tooltip: string,
+
+    /**
+     * From which direction the tooltip should appear, relative to the
+     * item. Used only on web.
+     */
+    tooltipPosition: string,
+
+    /**
+     * Whether this item is visible or not.
+     */
+    visible: boolean
+};
+
+/**
+ * Abstract (base) class for an item in {@link Toolbox}. The item can be located
+ * anywhere in the {@link Toolbox}, it will morph its shape to accommodate it.
+ *
+ * @abstract
+ */
+export default class AbstractToolboxItem<P : Props> extends Component<P> {
```

Comment:
There is some weird Flux syntax in this PR, the amount of time I spent taming Flux is nontrivial.